src/rgw/rgw_acl_swift.cc

```diff
@@ -216,10 +219,44 @@ int RGWAccessControlPolicy_SWIFT::create(RGWRados* const store,
                     << r << dendl;
       return r;
     }
+    rw_mask |= SWIFT_PERM_WRITE;
   }
   return 0;
 }
 
+void RGWAccessControlPolicy_SWIFT::filter_merge(int rw_mask,
+                                                RGWAccessControlPolicy_SWIFT *old)
+{
+/* rw_mask&SWIFT_PERM_READ => setting read acl,
+ * rw_mask&SWIFT_PERM_WRITE => setting write acl
+ * if bit is cleared, then copy matching elements from old.
+ */
+  if (rw_mask == (SWIFT_PERM_READ|SWIFT_PERM_WRITE)) {
+    return;
+  }
+  rw_mask ^= (SWIFT_PERM_READ|SWIFT_PERM_WRITE);
+  multimap<string, ACLGrant> m = old->acl.get_grant_map();
+  multimap<string, ACLGrant>::iterator iter;
+  for (iter = m.begin(); iter != m.end(); ++iter) {
+    ACLGrant& grant = iter->second;
+    uint32_t perm = grant.get_permission().get_permissions();
+    rgw_user id;
+    string url_spec;
+    if (!grant.get_id(id)) {
+      if (grant.get_group() != ACL_GROUP_ALL_USERS) {
+        url_spec = grant.get_referer();
+        if (url_spec.empty()) {
+          continue;
+        }
+	if (perm == 0) perm = SWIFT_PERM_READ;
+      }
+    }
+    if (perm & rw_mask) {
```

Comment:
As the `rw_mask` bases on general perm concept (`SWIFT_PERM_READ`, `SWIFT_PERM_WRITE`) and cooperates with the `perm` variable here, I believe it should be typed as `uint32_t`.